lib/Optimizer/GraphOptimizer/NodeSplitting.cpp

```diff
@@ -0,0 +1,1560 @@
+    clone->setName(node->getName().str() + ".Split" + std::to_string(sliceIdx));
+
+    // Gather final input slice ranges for the clone.
+    std::vector<SliceRange> inputRanges(clone->getNumInputs());
+    for (const auto &inputIdxMap : inputIdxAndMaps) {
+      auto inputCheckedRange = inputIdxMap.second(splitOutputSlice);
+      inputRanges[inputIdxMap.first] = inputCheckedRange.second;
+    }
+
+    // Gather final output slice ranges for the clone.
+    std::vector<SliceRange> outputRanges(clone->getNumResults());
+    outputRanges[splitOutputIdx] = splitOutputSlice;
+    for (const auto &outputIdxMap : outputIdxAndMaps) {
+      auto outputCheckedRange = outputIdxMap.second(splitOutputSlice);
+      outputRanges[outputIdxMap.first] = outputCheckedRange.second;
+    }
+
+    // Create input Slice nodes.
+    for (const auto &inputIdxMap : inputIdxAndMaps) {
+      auto inputIdx = inputIdxMap.first;
+      auto &inputRange = inputRanges[inputIdxMap.first];
+      Type outTy = Type::newShape(*(node->getNthInput(inputIdx).getType()),
+                                  inputRange.getSizes());
+      auto nodeName =
+          clone->getName().str() + ".SliceInput" + std::to_string(inputIdx);
+      auto *inputSlice = F->createSlice(nodeName, node->getNthInput(inputIdx),
+                                        inputRange.getStarts(), &outTy);
+      clone->setNthInput(inputIdx, inputSlice);
+    }
+
+    // Set clone split output type. The original node output type is not
+    // modified because the clone owns its output types.
+    TypeRef splitOutputType = F->getParent()->uniqueTypeWithNewShape(
+        node->getType(splitOutputIdx), splitOutputSlice.getSizes());
+    clone->getNthResult(splitOutputIdx).setTypeUnsafe(splitOutputType);
+
+    // Set clone output types. The original node output types are not
+    // modified because the clone owns its output types.
+    for (const auto &outputIdxMap : outputIdxAndMaps) {
+      auto &outputRange = outputRanges[outputIdxMap.first];
+      TypeRef outputType = F->getParent()->uniqueTypeWithNewShape(
+          node->getType(outputIdxMap.first), outputRange.getSizes());
+      clone->getNthResult(outputIdxMap.first).setTypeUnsafe(outputType);
+    }
+
+    // Modify clone.
+    splitNodeModifier(node, clone, inputRanges, outputRanges);
+
+    // Verify clone.
+    RETURN_ERR_IF_NOT(clone->verify(),
+                      "Invalid node obtained during node splitting!");
+
+    // Add clone to function.
+    F->addNode(clone);
+
+    // Add clone to vector.
+    splitNodes[sliceIdx] = clone;
+
+    // Merge the partial outputs of this clone.
+    for (size_t outIdx = 0; outIdx < node->getNumResults(); outIdx++) {
+      auto nodeName =
+          clone->getName().str() + ".MergeOutput" + std::to_string(outIdx);
+      mergedOutputs[outIdx] = F->createInsertTensor(
+          nodeName, mergedOutputs[outIdx], clone->getNthResult(outIdx),
+          outputRanges[outIdx].getStarts());
```

Comment:
Not quite.
The ConcatNode has the limitation of concatenating tensors along a single axis. The slices I`m dealing with here are generally split along multiple axes/dimensions. This functionality is achieved with InsertTensor nodes which are generic enough. In principle this could get achieved also with Concat nodes but would have a slightly more complex topology (multiple layers of Concats for each axis which get finally concatenated for the final result).
Another thing I don't like about Concat is that it forces/imposes the order in which the slices are concatenated/merged.
Having said that, I plan to leave this logic with InsertTensor nodes which are more flexible. A follow-up of this PR will be for me to add a good `EliminateInsertSlice` optimization pass.

Reply:
Ok, I guess it's more reasonable since you may be splitting on multiple axes. I still think it's doable to have a Concat per axis, and you'd be able to take advantage of concat optimizations. But it's not a huge deal.

Reply:
I would leave it as is and have a follow-up PR with the `EliminateInsertSlice` optimization pass.